### DISCUSSION TITLE
How to reduce RAM utilization in FINN?

### DISCUSSION THREAD

**Discussion by MFJI** [Xilinx/finn#1071] [Q&A] [answered]
Hello,

Anybody knows what can be the ways to reduce RAM utilization in FINN implementation?

Thanks.

**Comment by Ba1tu3han** [comment]
> Hello,
> 
> 
> 
> Anybody knows what can be the ways to reduce RAM utilization in FINN implementation?
> 
> 
> 
> Thanks.

Hello,

You should modify the folding configuration file (PE and SIMD parameters). As reference, look at the folding notebook in advance notebooks.

Batuhan

**Reply by MFJI** [reply]
Hello Batuhan,

Thank you for your answer. When the parallelism is already the lowest, is there still a way to reduce the RAM utilization?

**Reply by Ba1tu3han** [reply]
Indeed this kind of modifications have limits. You can try to use DSP and LUT instead of RAM by configuration file parameters. I can recommend that you should try to reduce number of parameters in your model. It effects so much.

**Reply by MFJI** [reply]
Thank you for your answer. What is the way to define which resource (LUT, BRAM, URAM?) to use as RAM in FINN?

**Reply by Ba1tu3han** [reply, marked as answer]
You should check this notebook and change ram_style as auto, distributed, block or ultra in custom configuration file that is created by yourself. https://github.com/Xilinx/finn/blob/e3087ad9fbabcc35f21164d415ababec4f462e9f/notebooks/advanced/4_advanced_builder_settings.ipynb#L987 

Please check this notebook for the folding filehttps://github.com/Xilinx/finn/blob/e3087ad9fbabcc35f21164d415ababec4f462e9f/notebooks/advanced/3_folding.ipynb


![image](https://github.com/Xilinx/finn/assets/29502318/a645d4e1-705e-4dde-b973-f119cb792954)

**Reply by MFJI** [reply]
Thank you so much
